raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftMetricsTest.java

```diff
@@ -0,0 +1,255 @@
+        assertEquals((double) 0, getMetric(metrics, "log-end-epoch").metricValue());
+
+        raftMetrics.updateLogEnd(new OffsetAndEpoch(5L, 1));
+
+        assertEquals((double) 5L, getMetric(metrics, "log-end-offset").metricValue());
+        assertEquals((double) 1, getMetric(metrics, "log-end-epoch").metricValue());
+    }
+
+    @Test
+    public void shouldRecordNumUnknownVoterConnections() throws IOException {
+        state = new QuorumState(localId, Collections.singleton(localId), new MockQuorumStateStore(), new LogContext("kafka-raft-metrics-test"));
+        state.initialize(new OffsetAndEpoch(0L, 0));
+        raftMetrics = new KafkaRaftMetrics(metrics, "raft", state, time.milliseconds());
+
+        assertEquals((double) 0, getMetric(metrics, "number-unknown-voter-connections").metricValue());
+
+        raftMetrics.updateNumUnknownVoterConnections(2);
+
+        assertEquals((double) 2, getMetric(metrics, "number-unknown-voter-connections").metricValue());
+    }
+
+    @Test
+    public void shouldRecordPollIdleRatio() throws IOException {
+        state = new QuorumState(localId, Collections.singleton(localId), new MockQuorumStateStore(), new LogContext("kafka-raft-metrics-test"));
+        state.initialize(new OffsetAndEpoch(0L, 0));
+        raftMetrics = new KafkaRaftMetrics(metrics, "raft", state, time.milliseconds());
+
+        raftMetrics.updatePollStart(time.milliseconds());
+        time.sleep(100L);
+        raftMetrics.updatePollEnd(time.milliseconds());
+        time.sleep(900L);
+        raftMetrics.updatePollStart(time.milliseconds());
+
+        assertEquals(0.1, getMetric(metrics, "poll-idle-ratio-avg").metricValue());
+
+        time.sleep(100L);
+        raftMetrics.updatePollEnd(time.milliseconds());
+        time.sleep(100L);
+        raftMetrics.updatePollStart(time.milliseconds());
+
+        assertEquals(0.3, getMetric(metrics, "poll-idle-ratio-avg").metricValue());
+    }
+
+    @Test
+    public void shouldRecordLatency() throws IOException {
+        state = new QuorumState(localId, Collections.singleton(localId), new MockQuorumStateStore(), new LogContext("kafka-raft-metrics-test"));
+        state.initialize(new OffsetAndEpoch(0L, 0));
+        raftMetrics = new KafkaRaftMetrics(metrics, "raft", state, time.milliseconds());
+
+        raftMetrics.updateElectionLatency(1000, time.milliseconds());
+
+        assertEquals((double) 1000, getMetric(metrics, "election-latency-avg").metricValue());
+        assertEquals((double) 1000, getMetric(metrics, "election-latency-max").metricValue());
+
+        raftMetrics.updateElectionLatency(800, time.milliseconds());
+
+        assertEquals((double) 900, getMetric(metrics, "election-latency-avg").metricValue());
+        assertEquals((double) 1000, getMetric(metrics, "election-latency-max").metricValue());
+
+        raftMetrics.updateCommitLatency(50.5, time.milliseconds());
```

Comment:
nit: unless .5 is necessary, let's use numbers that are easier to add and divide, like 40.0 and 50.0

Reply:
I used .5 intentionally but I think you have a good point here that JVM may grok about it, will change.